README.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> ElasticSearch components for Pip.Services in Node.js / ES2017

This module is a part of the [Pip.Services](http://pipservices.org) polyglot microservices toolkit.

The Elasticsearch module contains logging components with data storage on the Elasticsearch server.

The module contains the following packages:
- **Build** - contains a factory for the construction of components
- **Log** - Logging components

<a name="links"></a> Quick links:

* [Configuration](https://www.pipservices.org/recipies/configuration)
* [Logging](https://www.pipservices.org/recipies/active-logic)
* [Virtual memory configuration](https://www.elastic.co/guide/en/elasticsearch/reference/current/vm-max-map-count.html)
* [API Reference](https://pip-services3-nodex.github.io/pip-services3-elasticsearch-nodex/globals.html)
* [Change Log](CHANGELOG.md)
* [Get Help](https://www.pipservices.org/community/help)
* [Contribute](https://www.pipservices.org/community/contribute)

## Use

Install the NPM package as
```bash
npm install pip-services-elasticsearch-nodex --save
```

Microservice components shall perform logging usual way using CompositeLogger component.
The CompositeLogger will get ElasticSearchLogger from references and will redirect log messages
there among other destinations.

```typescript
import { ConfigParams } from 'pip-services3-commons-nodex'; 
import { IConfigurable } from 'pip-services3-commons-nodex'; 
import { IReferences } from 'pip-services3-commons-nodex'; 
import { IReferenceable } from 'pip-services3-commons-nodex'; 
import { CompositeLogger } from 'pip-services3-components-nodex'; 

export class MyComponent implements IConfigurable, IReferenceable {
  private _logger: CompositeLogger = new CompositeLogger();
  
  public configure(config: ConfigParams): void {
    this._logger.configure(config);
  }
  
  public setReferences(refs: IReferences): void {
    this._logger.setReferences(refs);
  }
  
  public myMethod(correlationId: string, param1: any, callback: (err: any, result: any) => void): void {
    this._logger.trace(correlationId, "Executed method mycomponent.mymethod");
    ....
  }
}
```

Configuration for your microservice that includes ElasticSearch logger may look the following way.

```yaml
...
{{#if ELASTICSEARCH_ENABLED}}
- descriptor: pip-services:logger:elasticsearch:default:1.0
  connection:
    uri: {{{ELASTICSEARCG_SERVICE_URI}}}
    host: {{{ELASTICSEARCH_SERVICE_HOST}}}{{#unless ELASTICSEARCH_SERVICE_HOST}}localhost{{/unless}}
    port: {{ELASTICSEARCG_SERVICE_PORT}}{{#unless ELASTICSEARCH_SERVICE_PORT}}9200{{/unless}}\ 
{{/if}}
...
```

## Develop

For development you shall install the following prerequisites:
* Node.js 8+
* Visual Studio Code or another IDE of your choice
* Docker
* Typescript

Install dependencies:
```bash
npm install
```

Compile the code:
```bash
tsc
```

Run automated tests:
```bash
npm test
```

Generate API documentation:
```bash
./docgen.ps1
```

Before committing changes run dockerized build and test as:
```bash
./build.ps1
./test.ps1
./clear.ps1
```

Configure the vm.max_map_count

`sudo sysctl -w vm.max_map_count=262144`

    fixes:
    max virtual memory areas vm.max_map_count [65530] is too low, increase to at least [262144]

## Typed vs Type-less indexes
ES version 7 [stopped supporting "types"](https://www.elastic.co/guide/en/elasticsearch/reference/current/removal-of-types.html) and encouraged a separation of disperate data into different indexes. By default, this version of pip-services3-elasticsearch-nodex will support 7.x type-less indexes. You can support to the 6.x "typed" approach by setting the `include_type_name` option to true. This allows it to work with either 6.x or 7.x ElasticSearch servers.

You can read more about how this is accomplished  [here](https://www.elastic.co/blog/moving-from-types-to-typeless-apis-in-elasticsearch-7-0)

## Contacts

The library is created and maintained by **Sergey Seroukhov**.

The documentation is written by:
- **Mark Makarychev**




CHANGELOG.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> ElasticSearch components for Pip.Services in Node.js / ES2017 Changelog

## <a name="1.0.0"></a> 1.0.0 (2021-05-28) 

Initial public release

### Features
* Added DefaultElasticSearchFactory
* Added ElasticSearchLogger


test/fixtures/LoggerFixture.ts
let assert = require('chai').assert;

import { LogLevel } from 'pip-services3-components-nodex';
import { CachedLogger } from 'pip-services3-components-nodex';

export class LoggerFixture {
    private _logger: CachedLogger;

    public constructor(logger: CachedLogger) {
        this._logger = logger;
    }

    public testLogLevel() {
        assert.isTrue(this._logger.getLevel() >= LogLevel.None);
        assert.isTrue(this._logger.getLevel() <= LogLevel.Trace);
    }

    public async testSimpleLogging() {
        this._logger.setLevel(LogLevel.Trace);

        this._logger.fatal(null, null, "Fatal error message");
        this._logger.error(null, null, "Error message");
        this._logger.warn(null, "Warning message");
        this._logger.info(null, "Information message");
        this._logger.debug(null, "Debug message");
        this._logger.trace(null, "Trace message");

        this._logger.dump();
        
        await new Promise<void>((resolve, reject) => { setTimeout(resolve, 1000); });
    }

    public async testErrorLogging() {
        try {
            // Raise an exception
            throw new Error();
        } catch (ex) {
            this._logger.fatal("123", ex, "Fatal error");
            this._logger.error("123", ex, "Recoverable error");

            assert.isNotNull(ex);
        }

        this._logger.dump();

        await new Promise<void>((resolve, reject) => { setTimeout(resolve, 1000); });
    }
    
}

test/log/ElasticSearchLogger.test.ts
import { ConfigParams } from 'pip-services3-commons-nodex';

import { ElasticSearchLogger } from '../../src/log/ElasticSearchLogger';
import { LoggerFixture } from '../fixtures/LoggerFixture';

let assert = require('chai').assert;

suite('ElasticSearchLogger', () => {
    let _logger: ElasticSearchLogger;
    let _fixture: LoggerFixture;

    setup(async () => {
        let host = process.env['ELASTICSEARCH_SERVICE_HOST'] || 'localhost';
        let port = process.env['ELASTICSEARCH_SERVICE_PORT'] || 9200;
        let dateFormat: string = "YYYYMMDD";

        _logger = new ElasticSearchLogger();
        _fixture = new LoggerFixture(_logger);

        let config = ConfigParams.fromTuples(
            'source', 'test',
            'index', 'log',
            'daily', true,
            "date_format", dateFormat,
            'connection.host', host,
            'connection.port', port
        );
        _logger.configure(config);

        await _logger.open(null);
    });

    teardown(async () => {
        await _logger.close(null);
    });

    test('Log Level', () => {
        _fixture.testLogLevel();
    });

    test('Simple Logging', async () => {
        await _fixture.testSimpleLogging();
    });

    test('Error Logging', async () => {
        await _fixture.testErrorLogging();
    });

    /**
     * We test to ensure that the date pattern does not effect the opening of the elasticsearch component
     */
    test('Date Pattern Testing - YYYY.MM.DD', async () => {
        let host = process.env['ELASTICSEARCH_SERVICE_HOST'] || 'localhost';
        let port = process.env['ELASTICSEARCH_SERVICE_PORT'] || 9200;

        let logger = new ElasticSearchLogger();
        let dateFormat: string = "YYYY.MM.DD";

        let config = ConfigParams.fromTuples(
            'source', 'test',
            'index', 'log',
            'daily', true,
            "date_format", dateFormat,
            'connection.host', host,
            'connection.port', port
        );
        logger.configure(config);

        await logger.open(null);

        // Since the currentIndex property is private, we will just check for an open connection
        assert.isTrue(logger.isOpen());
        await logger.close(null);
    });

    /**
     * We test to ensure that the date pattern does not effect the opening of the elasticsearch component
     */
    test('Date Pattern Testing - YYYY.M.DD', async () => {
        let host = process.env['ELASTICSEARCH_SERVICE_HOST'] || 'localhost';
        let port = process.env['ELASTICSEARCH_SERVICE_PORT'] || 9200;

        let logger = new ElasticSearchLogger();
        let dateFormat: string = "YYYY.M.DD";

        let config = ConfigParams.fromTuples(
            'source', 'test',
            'index', 'log',
            'daily', true,
            "date_format", dateFormat,
            'connection.host', host,
            'connection.port', port
        );
        logger.configure(config);

        await logger.open(null);

        // Since the currentIndex property is private, we will just check for an open connection
        assert.isTrue(logger.isOpen());
        await logger.close(null);
    });

});

src/log/ElasticSearchLogger.ts
/** @module log */
import * as moment from 'moment';

import { ConfigParams } from 'pip-services3-commons-nodex';
import { IReferences } from 'pip-services3-commons-nodex';
import { IReferenceable } from 'pip-services3-commons-nodex';
import { IOpenable } from 'pip-services3-commons-nodex';
import { IdGenerator } from 'pip-services3-commons-nodex';
import { HttpConnectionResolver } from 'pip-services3-rpc-nodex';
import { ConfigException } from 'pip-services3-commons-nodex';
import { CachedLogger } from 'pip-services3-components-nodex';
import { LogMessage } from 'pip-services3-components-nodex';

/**
 * Logger that dumps execution logs to ElasticSearch service.
 * 
 * ElasticSearch is a popular search index. It is often used 
 * to store and index execution logs by itself or as a part of
 * ELK (ElasticSearch - Logstash - Kibana) stack.
 * 
 * Authentication is not supported in this version.
 * 
 * ### Configuration parameters ###
 * 
 * - level:             maximum log level to capture
 * - source:            source (context) name
 * - connection(s):
 *     - discovery_key:         (optional) a key to retrieve the connection from [[https://pip-services3-nodex.github.io/pip-services3-components-nodex/interfaces/connect.idiscovery.html IDiscovery]]
 *     - protocol:              connection protocol: http or https
 *     - host:                  host name or IP address
 *     - port:                  port number
 *     - uri:                   resource URI or connection string with all parameters in it
 * - options:
 *     - interval:          interval in milliseconds to save log messages (default: 10 seconds)
 *     - max_cache_size:    maximum number of messages stored in this cache (default: 100)
 *     - index:             ElasticSearch index name (default: "log")
 *     - date_format        The date format to use when creating the index name. Eg. log-YYYYMMDD (default: "YYYYMMDD"). See [[https://momentjs.com/docs/#/displaying/format/ Moment.js]]
 *     - daily:             true to create a new index every day by adding date suffix to the index name (default: false)
 *     - reconnect:         reconnect timeout in milliseconds (default: 60 sec)
 *     - timeout:           invocation timeout in milliseconds (default: 30 sec)
 *     - max_retries:       maximum number of retries (default: 3)
 *     - index_message:     true to enable indexing for message object (default: false)
 *     - include_type_name: Will create using a "typed" index compatible with ElasticSearch 6.x (default: false)
 * 
 * ### References ###
 * 
 * - <code>\*:context-info:\*:\*:1.0</code>      (optional) [[https://pip-services3-nodex.github.io/pip-services3-components-nodex/classes/info.contextinfo.html ContextInfo]] to detect the context id and specify counters source
 * - <code>\*:discovery:\*:\*:1.0</code>         (optional) [[https://pip-services3-nodex.github.io/pip-services3-components-nodex/interfaces/connect.idiscovery.html IDiscovery]] services to resolve connection
 * 
 * ### Example ###
 * 
 *     let logger = new ElasticSearchLogger();
 *     logger.configure(ConfigParams.fromTuples(
 *         "connection.protocol", "http",
 *         "connection.host", "localhost",
 *         "connection.port", 9200
 *     ));
 *     
 *     await logger.open("123");
 *     
 *     logger.error("123", ex, "Error occured: %s", ex.message);
 *     logger.debug("123", "Everything is OK.");
 */
export class ElasticSearchLogger extends CachedLogger implements IReferenceable, IOpenable {
    private _connectionResolver: HttpConnectionResolver = new HttpConnectionResolver();

    private _timer: any;
    private _index: string = "log";
    private _dateFormat: string = "YYYYMMDD";
    private _dailyIndex: boolean = false;
    private _currentIndex: string;
    private _reconnect: number = 60000;
    private _timeout: number = 30000;
    private _maxRetries: number = 3;
    private _indexMessage: boolean = false;
    private _include_type_name: boolean = false;
    private _client: any = null;

    /**
     * Creates a new instance of the logger.
     */
    public constructor() {
        super();
    }

    /**
     * Configures component by passing configuration parameters.
     * 
     * @param config    configuration parameters to be set.
     */
    public configure(config: ConfigParams): void {
        super.configure(config);

        this._connectionResolver.configure(config);

        this._index = config.getAsStringWithDefault('index', this._index);
        this._dateFormat = config.getAsStringWithDefault("date_format", this._dateFormat);
        this._dailyIndex = config.getAsBooleanWithDefault('daily', this._dailyIndex);
        this._reconnect = config.getAsIntegerWithDefault('options.reconnect', this._reconnect);
        this._timeout = config.getAsIntegerWithDefault('options.timeout', this._timeout);
        this._maxRetries = config.getAsIntegerWithDefault('options.max_retries', this._maxRetries);
        this._indexMessage = config.getAsBooleanWithDefault('options.index_message', this._indexMessage);
        this._include_type_name = config.getAsBooleanWithDefault('options.include_type_name', this._include_type_name);
    }

    /**
     * Sets references to dependent components.
     * 
     * @param references 	references to locate the component dependencies. 
     */
    public setReferences(references: IReferences): void {
        super.setReferences(references);
        this._connectionResolver.setReferences(references);
    }

    /**
     * Checks if the component is opened.
     * 
     * @returns true if the component has been opened and false otherwise.
     */
    public isOpen(): boolean {
        return this._timer != null;
    }

    /**
     * Opens the component.
     * 
     * @param correlationId 	(optional) transaction id to trace execution through call chain.
     */
    public async open(correlationId: string): Promise<void> {
        if (this.isOpen()) {
            return;
        }

        let connection = await this._connectionResolver.resolve(correlationId);
        if (connection == null) {
            throw new ConfigException(
                correlationId,
                'NO_CONNECTION',
                'Connection is not configured'
            );
        }

        let uri = connection.getAsString("uri");

        let options = {
            host: uri,
            requestTimeout: this._timeout,
            deadTimeout: this._reconnect,
            maxRetries: this._maxRetries
        };

        let elasticsearch = require('elasticsearch');
        this._client = new elasticsearch.Client(options);

        await this.createIndexIfNeeded(correlationId, true);

        this._timer = setInterval(() => { this.dump() }, this._interval);
    }

    /**
     * Closes component and frees used resources.
     * 
     * @param correlationId 	(optional) transaction id to trace execution through call chain.
     */
    public async close(correlationId: string): Promise<void> {
        await this.save(this._cache);

        if (this._timer) {
            clearInterval(this._timer);
        }

        this._cache = [];
        this._timer = null;
        this._client = null;
    }

    private getCurrentIndex(): string {
        if (!this._dailyIndex) return this._index;

        let today = new Date().toUTCString();
        let datePattern = moment(today).format(this._dateFormat);

        return this._index + "-" + datePattern;
    }

    private async createIndexIfNeeded(correlationId: string, force: boolean): Promise<void> {
        let newIndex = this.getCurrentIndex();
        if (!force && this._currentIndex == newIndex) {
            return;
        }

        this._currentIndex = newIndex;

        let exists = new Promise<boolean>((resolve, reject) => {
            this._client.indices.exists(
                { index: this._currentIndex },
                (err, exists) => {
                    if (err != null) {
                        reject(err);
                        return;
                    }
                    resolve(exists);
                }
            );    
        });

        if (exists) {
            return;
        }

        await new Promise<void>((resolve, reject) => {
            this._client.indices.create(
                {
                    include_type_name: this._include_type_name,
                    index: this._currentIndex,
                    body: {
                        settings: {
                            number_of_shards: 1
                        },
                        mappings: this.getIndexSchema()
                    }
                },
                (err) => {
                    // Skip already exist errors
                    if (err != null && err.message.indexOf('resource_already_exists') >= 0) {
                        err = null;
                    }

                    if (err != null) {
                        reject(err);
                        return;
                    }

                    resolve();
                }
            );
        });
    }

    /**
     * Returns the schema of the log message
     * @param include_type_name A flag that indicates whether the schema should follow the pre-ES 6.x convention
     */
    private getIndexSchema(): any {

        const schema = {
            properties: {
                time: { type: "date", index: true },
                source: { type: "keyword", index: true },
                level: { type: "keyword", index: true },
                correlation_id: { type: "text", index: true },
                error: {
                    type: "object",
                    properties: {
                        type: { type: "keyword", index: true },
                        category: { type: "keyword", index: true },
                        status: { type: "integer", index: false },
                        code: { type: "keyword", index: true },
                        message: { type: "text", index: false },
                        details: { type: "object" },
                        correlation_id: { type: "text", index: false },
                        cause: { type: "text", index: false },
                        stack_trace: { type: "text", index: false }
                    }
                },
                message: { type: "text", index: this._indexMessage }
            }
        }

        if (this._include_type_name) {
            return {
                log_message: schema
            }
        } else return schema
    }

    /**
     * Saves log messages from the cache.
     * 
     * @param messages  a list with log messages
     */
    protected async save(messages: LogMessage[]): Promise<void> {
        if (!this.isOpen() || messages.length == 0) {
            return;
        }

        await this.createIndexIfNeeded('elasticsearch_logger', false);

        let bulk = [];
        for (let message of messages) {
            bulk.push({ index: this.getLogItem()})
            bulk.push(message);
        }

        await new Promise<void>((resolve, reject) => {
            this._client.bulk({ body: bulk }, (err) => {
                if (err != null) {
                    reject(err);
                    return;
                }
                resolve();
            });
        });
    }

    protected getLogItem(): any {
        return this._include_type_name ?
            { _index: this._currentIndex, _type: "log_message", _id: IdGenerator.nextLong() } : // ElasticSearch 6.x
            { _index: this._currentIndex, _id: IdGenerator.nextLong() }                         // ElasticSearch 7.x
    }
}